### PR TITLE
QF-1571_Withdrawal_journey_changes

### PR DESCRIPTION
**Pull request by ChrisJWoodcock**
…sion.

### CODE REVIEW

File: src/SFA.DAS.AdminService.Web/Services/AnswerInjectionService.cs [hunk 1]
```diff
@@ -229,31 +229,29 @@ public async Task<CreateOrganisationStandardFromApplyResponse> InjectApplyOrgani
...
             if (warningMessages.Count == 0)
             {
                 var validationResponse = await _assessorValidationService.ValidateNewOrganisationStandardRequest(standard);
 
-                if (standard.ApplyFollowingWithdrawal && validationResponse.Errors.Count == 1 &&
+                // When applying for a standard that was previously withdrawn
+                var applyFollowingWithdrawal = await _applyApiClient.GetLatestWithdrawalDateForStandard(command.OrganisationId, command.StandardCode) != null ? true : false;
+
+                // The organisation is expected to be present already
+                if (applyFollowingWithdrawal && validationResponse.Errors.Count == 1 &&
```
Comment (SawNawDfE) [inline src/SFA.DAS.AdminService.Web/Services/AnswerInjectionService.cs:250]:
```suggestion
                if (applyFollowingWithdrawal && validationResponse.Errors.Single() &&
```

File: src/SFA.DAS.AdminService.Web/Services/AnswerService.cs [hunk 2]
```diff
@@ -164,6 +166,39 @@ public async Task<CreateOrganisationStandardCommand> GatherAnswersForOrganisatio
...
+            };
+        }
+
+        public async Task<WithdrawStandardRequest> GatherAnswersForWithdrawStandardForApplication(Guid applicationId)
+        {
+            var application = await _applyApiClient.GetApplication(applicationId);
+            var applicationData = await _qnaApiClient.GetApplicationDataDictionary(application?.ApplicationId ?? Guid.Empty);
+            var organisation = await _apiApiClient.GetOrganisation(application?.OrganisationId ?? Guid.Empty);
+
+            if (application is null || applicationData is null || organisation is null) return null;
```
Comment (SawNawDfE) [inline src/SFA.DAS.AdminService.Web/Services/AnswerService.cs:192]: These lines are a duplicate of lines 171-175. If this is intentional, they can be refactored into a class that returns the three as a DTO. That way we can add some quick tests. (Test coverage is very low in this PR and in the Admin Service in general)

**Review by SawNawDfE — CHANGES_REQUESTED** [review]
Left a few comments

File: src/SFA.DAS.AdminService.Web/Services/AnswerInjectionService.cs (hunk 1 above)
Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/Services/AnswerInjectionService.cs:250]: That suggestion is not equivalent .Single() will throw an exception if there is not 1 item - this check is checking if there is 1 item

File: src/SFA.DAS.AdminService.Web/Services/ApplicationService.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        private readonly IQnaApiClient _qnaApiClient;
+
+        public ApplicationService(IApiClient apiClient, IApplicationApiClient applyApiClient, IQnaApiClient qnaApiClient)
+        {
+            _apiApiClient = apiClient;
+            _applyApiClient = applyApiClient;
+            _qnaApiClient = qnaApiClient;
+        }
+
+        public async Task<WithdrawalApplicationDetails> GetWithdrawalApplicationDetails(Guid applicationId)
```
Comment (SawNawDfE) [inline src/SFA.DAS.AdminService.Web/Services/ApplicationService.cs:24]: The return type should indicate that a null return is possible. The compiler then marks the calling line as "maybe null", and code analysis tools can make use of this
```suggestion
        public async Task<WithdrawalApplicationDetails?> GetWithdrawalApplicationDetails(Guid applicationId)
```

File: src/SFA.DAS.AdminService.Web/Services/ApplicationService.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            return new WithdrawalApplicationDetails
+            {
+                ApplicationId = application.Id,
+                EndPointAssessorOrganisationId = organisation.EndPointAssessorOrganisationId,
+                StandardCode = application.StandardCode.Value,
+                ConfirmedWithdrawalDate = (DateTime)applicationData[nameof(ApplicationData.ConfirmedWithdrawalDate)]
+            };
+        }
+
+        public async Task<ApplicationDetails> GetApplicationsDetails(Guid applicationId)
```
Comment (SawNawDfE) [inline src/SFA.DAS.AdminService.Web/Services/ApplicationService.cs:41]: Same here
```suggestion
        public async Task<ApplicationDetails?> GetApplicationsDetails(Guid applicationId)
```

**Review by SawNawDfE — CHANGES_REQUESTED** [review]

**Review by SawNawDfE — DISMISSED** [review]

File: src/SFA.DAS.AdminService.Web/Services/ApplicationService.cs (hunk 3 above)
Comment (SawNawDfE) [reply src/SFA.DAS.AdminService.Web/Services/ApplicationService.cs:24]: I just noticed that a few changes to the project file would be needed to make use of nullables. If you decide to do this, you would have to add these lines to the PropertyGroup section of the csproj:

```
<LangVersion>10.0</LangVersion> <!-- 8.0 or higher will work, but .NET 6 supports C# 10-->
<Nullable>enable</Nullable>
```

This should've been done as part of the .NET 6 upgrade. It's unfortunate that our .NET 6 apps are still using the version of C# released with .NET Core 2.0!

Comment (ChrisJWoodcock) [reply src/SFA.DAS.AdminService.Web/Services/ApplicationService.cs:24]: I'll leave these for now then

**Review by SawNawDfE — DISMISSED** [review]

**Review by SawNawDfE — APPROVED** [review]

**Review by furqanagwan — DISMISSED** [review]

**Review by chrisfoster76 — DISMISSED** [review]

**Review by debrr07 — APPROVED** [review]

**Pull request merged**